src/libraries/System.Net.Http/src/System/Net/Http/Metrics/HttpMetricsEnrichmentContext.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Net.Http.Metrics;
+
+public class HttpMetricsEnrichmentContext
+{
+    private HttpRequestMessage? _request;
+    private HttpResponseMessage? _response;
+    private Exception? _exception;
+    private TagCollection _tags;
+
+    public HttpMetricsEnrichmentContext()
+    {
+        _tags = new TagCollection();
+    }
+
+    public HttpRequestMessage Request
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _request!;
+        }
+    }
+
+    public HttpResponseMessage? Response
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _response;
+        }
+    }
+
+    public Exception? Exception
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _exception;
+        }
+    }
+
+    internal bool InProgress => _request != null;
+
+    public ICollection<KeyValuePair<string, object?>> CustomTags => _tags;
+
+    internal void ApplyEnrichment(HttpRequestMessage request, HttpResponseMessage? response, Exception? exception, ref TagList tags)
+    {
+        if (request._options?.TryGetMetricsEnrichmentCallbackCollection(out List<Action<HttpMetricsEnrichmentContext>>? callbackCollection) != true)
+        {
+            return;
+        }
+
+        _request = request;
+        _response = response;
+        _exception = exception;
+        _tags._tags.Clear();
+        _tags._runningCallback = true;
+
+        try
+        {
+            foreach (Action<HttpMetricsEnrichmentContext> callback in callbackCollection!)
+            {
+                callback(this);
+            }
+        }
+        finally
+        {
+            _tags._runningCallback = false;
+            _request = null;
+            _response = null;
+            _exception = null;
+        }
+        ref TagList enrichmentTags = ref _tags._tags;
+        if (enrichmentTags.Count > 0)
```

Comment:
Unless TagList.Count is expensive, we needn't guard the for loop with this check.  And if it is expensive, we shouldn't be accessing it on every iteration of the loop.